src/metabase/query_processor/pivot/postprocess.clj

```diff
@@ -99,16 +100,28 @@
 (mu/defn init-pivot
   "Initiate the pivot data structure."
   [pivot-spec :- ::pivot-spec]
-  (let [{:keys [pivot-rows pivot-cols pivot-measures]} pivot-spec]
+  (let [{:keys [pivot-measures]} pivot-spec]
     {:config         pivot-spec
      :data           {}
-     :row-values     (zipmap pivot-rows (repeat (sorted-set)))
-     :column-values  (zipmap pivot-cols (repeat (sorted-set)))
+     ;; A nested tree of ordered maps & sets, representing all combinations of row values in the data
+     :row-paths      (ordered-map/ordered-map)
+     ;; A nested tree of ordered maps & sets, representing all combinations of column values in the data
+     :col-paths      (ordered-map/ordered-map)
      :measure-values (zipmap pivot-measures (repeat (sorted-set)))}))
 
-(defn- update-set
-  [m k v]
-  (update m k conj v))
+(defn- assoc-in-path-tree
+  "Assocs a value in a path tree, which should consist of a hierarchy of ordered-maps, with leaf values stored in
+  ordered-sets."
+  [tree ks v]
+  (let [step (fn step [tree [k & ks]]
+               (if ks
+                 (if-let [next-map (get tree k)]
+                   (assoc tree k (step next-map ks))
+                   (assoc tree k (step (ordered-map/ordered-map) ks)))
+                 (if k
+                   (assoc tree k (conj (get tree k (ordered-set/ordered-set)) v))
```

Comment:
You don't want to use the 3-arg `get` arity here because the default value is computed every time. `or` is better here.